### PR TITLE
Download Strategy Dependency Missing Case

### PR DESCRIPTION
**Pull request by TheMDev**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

```
~ % brew style    
Inspecting 1169 files
.................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

1169 files inspected, no offenses detected
~ % brew typecheck
No errors! Great job.
~ % brew tests    
Randomized with seed 35402
16 processes for 399 specs, ~ 24 specs per process
.....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
.........................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
..........................................................................................................................................................................................................................................................................................................................................................................................................................................................
..............................................................................................................................................................................................................................................................................................................................
.....................................................
.................................................................................................................................
........................................................................................................................................................................................................................................................................
............................................................................
..........................................................................................................................................................................................................
.....................................................................
................................................................................................................................................................................
................................................................................................................................................
...........................................................................................
.................
..................................................................................................................................................................................................................................................................................................................................................
..................................................................................................................................................................
Took 240 seconds (4:00)
```
Fixing the issues discussed in https://github.com/Homebrew/brew/issues/15395 Summary:
Download Strategy :nounzip tries to resolve dependencies to extract the download even though they are unnecessary.
Download Dependency for rar files is broken due to legal issues.

### CODE REVIEW

**Comment by TheMDev** [comment]
I just noticed the unpack_strategy for rar files and the corresponding test. Should these be removed or modified, as well as they are likely broken?
Library/Homebrew/unpack_strategy/rar.rb
Library/Homebrew/test/unpack_strategy/rar_spec.rb

File: Library/Homebrew/dependency_collector.rb [hunk 1]
```diff
@@ -179,6 +179,8 @@ def resource_dep(spec, tags)
     if strategy <= HomebrewCurlDownloadStrategy
       @deps << curl_dep_if_needed(tags)
       parse_url_spec(spec.url, tags)
+    elsif strategy <= NoUnzipCurlDownloadStrategy
+      # allow no unzip to pass through
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dependency_collector.rb:183]:
```suggestion
      # ensure NoUnzip never adds any dependencies
```

**Review by MikeMcQuaid — COMMENTED** [review]
Looking great so far!

**Comment by MikeMcQuaid** [comment]
> I just noticed the unpack_strategy for rar files and the corresponding test. Should these be removed or modified, as well as they are likely broken?
> Library/Homebrew/unpack_strategy/rar.rb
> Library/Homebrew/test/unpack_strategy/rar_spec.rb

Ideally these would be modified to use an actual existing formula? Thoughts @Bo98?

**Comment by Bo98** [comment]
Yes. Feel free to leave the `unrar` bit untouched if you would prefer for me to look at it in a separate PR. The nounzip change looks good on its own.

File: Library/Homebrew/dependency_collector.rb (hunk 1 above)
Comment (reitermarkus) [reply Library/Homebrew/dependency_collector.rb:183]: This should be above the previous check.

**Review by reitermarkus — DISMISSED** [review]

**Comment by TheMDev** [comment]
@Bo98 
> Yes. Feel free to leave the `unrar` bit untouched if you would prefer for me to look at it in a separate PR. The nounzip change looks good on its own.

While I was waiting for the PR to get reviewed, I did a little bit more digging I am not sure if this would work, but I found cask dependent, which might be a drop-in replacement for Library/Homebrew/dependency_collector.rb:213? `CaskDependent.new("rar")`

And Library/Homebrew/unpack_strategy/rar.rb:19 should also be a one line fix with `@dependencies ||= [Cask::CaskLoader.load("rar")]`

Thoughts?

**Comment by apainintheneck** [comment]
Using a cask dependent here wouldn't work on Linux, right?

**Comment by MikeMcQuaid** [comment]
@apainintheneck Indeed and we don't like cask dependents for other reasons e.g. they may require root, they may not be open source.

**Comment by TheMDev** [comment]
@MikeMcQuaid I agree that the cask is not the optimal solution I just figured I would mention it as I noticed `Library/Homebrew/unpack_strategy/air.rb` uses a cask. Should I have `Library/Homebrew/unpack_strategy/rar.rb` throw an error like `Library/Homebrew/dependency_collector.rb` then?

**Comment by MikeMcQuaid** [comment]
@TheMDev If we're already using another cask I'd be potentially game for one here. What d you think @Homebrew/maintainers?

**Comment by Bo98** [comment]
I would like to know first what we gain from using the `rar` cask over `libarchive` or `unar`.

I recognise support might be better for cask version, but losing Linux support is also a big deal so there's going to be a balancing act of pros/cons. If there's little realistic support differences, we should use `libarchive` - or `unar` if that's better.

**Comment by Bo98** [comment]
> @MikeMcQuaid I agree that the cask is not the optimal solution I just figured I would mention it as I noticed `Library/Homebrew/unpack_strategy/air.rb` uses a cask. Should I have `Library/Homebrew/unpack_strategy/rar.rb` throw an error like `Library/Homebrew/dependency_collector.rb` then?

FWIW I consider Air archives to be extremely rare. I've never seen one used in practice for distributing files that need extracting.

So I'm not surprised no one has complained about no Linux support there.

RAR on the other hand is something I've dealt with on multiple occasions.

**Comment by TheMDev** [comment]
@Bo98 Sorry for the confusion; one of my posts may have gotten lost. unar only includes little of the functionality that libarchive does. I was originally against libarchive as well because it kept throwing errors while trying to extract. However, after further looking into it, that seems to have been a user error, and I just needed to read the caveats.
```
==> Caveats
libarchive is keg-only, which means it was not symlinked into /usr/local,
because macOS already provides this software and installing another version in
parallel can cause all kinds of trouble.

If you need to have libarchive first in your PATH, run:
  echo 'export PATH="/usr/local/opt/libarchive/bin:$PATH"' >> ~/.zshrc

For compilers to find libarchive you may need to set:
  export LDFLAGS="-L/usr/local/opt/libarchive/lib"
  export CPPFLAGS="-I/usr/local/opt/libarchive/include"

For pkg-config to find libarchive you may need to set:
  export PKG_CONFIG_PATH="/usr/local/opt/libarchive/lib/pkgconfig"
```

I think libarchive would be a much better solution than the cask. However, is there a way to specify a keg as the dependency? Do we want to future-proof this in some way so that if macOS does update to 3.6.0+, the keg is no longer needed as a dependency? If so, how would we go about that? Otherwise, I think its as simple as just calling `/usr/local/opt/libarchive/bin/bsdtar`

**Comment by Bo98** [comment]
> However, is there a way to specify a keg as the dependency?

Specify the dependency as a normal formula and if you adjust the env line as well to reference the correct formula:

https://github.com/Homebrew/brew/blob/f7b3225574c427da939ca32458004828a5be6523/Library/Homebrew/unpack_strategy/rar.rb#L28

it should work fine when calling `bsdtar` as it'll use that PATH.

> Do we want to future-proof this in some way so that if macOS does update to 3.6.0+, the keg is no longer needed as a dependency?

We don't know if or when Apple will update so really all to do is to leave a comment saying it can be adjusted when macOS ships with 3.6.0. I've also noted this down to check when macOS 14 betas start in a few weeks, alongwith the usual version change checks I do.

**Comment by TheMDev** [comment]
Okay, I have made the changes, reran the checks, ran the test formula, and everything seems to work now!
Thanks for the help.

```
~ % brew style    
Inspecting 1173 files
.....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

1173 files inspected, no offenses detected
~ % brew typecheck
No errors! Great job.
~ % brew tests    
Randomized with seed 49576
16 processes for 401 specs, ~ 25 specs per process
...............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
.......................................................................................................................................................................................................................................
.....................................
.......................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
....................
.............................................................
............................................................................................................................................................................................................................................................................................................................................................................................................
.........
...........................................................................................................................
.........................
...............................
..........................................................................................
.................................................
.................................................................................................................................................
...............................................................................................................................................................................................................................................................................................................................................
...........................................................................................................................
Took 216 seconds (3:36)
```

```
~ % HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source --verbose --debug test_rar
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/test_rar.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libarchive.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libb2.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/lz4.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/xz.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/zstd.rb
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/cmake.rb
==> Fetching test_rar
/usr/bin/env /usr/local/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.0.18-59-gf7b3225-dirty\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 13.3.1\)\ curl/7.87.0 --header Accept-Language:\ en --retry 3 --fail --location --silent --head https://archive.org/download/SpaceCadet_Plus95/Space_Cadet.rar
/usr/bin/env /usr/local/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.0.18-59-gf7b3225-dirty\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 13.3.1\)\ curl/7.87.0 --header Accept-Language:\ en --retry 3 --fail --location --silent --head --request GET https://archive.org/download/SpaceCadet_Plus95/Space_Cadet.rar
==> Downloading https://archive.org/download/SpaceCadet_Plus95/Space_Cadet.rar
==> Downloading from https://ia800601.us.archive.org/13/items/SpaceCadet_Plus95/Space_Cadet.rar
/usr/bin/env /usr/local/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.0.18-59-gf7b3225-dirty\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 13.3.1\)\ curl/7.87.0 --header Accept-Language:\ en --retry 3 --location --head https://ia800601.us.archive.org/13/items/SpaceCadet_Plus95/Space_Cadet.rar
/usr/bin/env /usr/local/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.0.18-59-gf7b3225-dirty\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 13.3.1\)\ curl/7.87.0 --header Accept-Language:\ en --fail --retry 3 --location --remote-time --output ~/Library/Caches/Homebrew/downloads/8c353ea6176f8b38777694436e18bb3d707b0586c77e38ea122558f22d80a3dc--Space_Cadet.rar.incomplete https://ia800601.us.archive.org/13/items/SpaceCadet_Plus95/Space_Cadet.rar
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1061k  100 1061k    0     0  1077k      0 --:--:-- --:--:-- --:--:-- 1084k
==> Verifying checksum for '8c353ea6176f8b38777694436e18bb3d707b0586c77e38ea122558f22d80a3dc--Space_Cadet.rar'
/usr/bin/env /usr/local/Homebrew/Library/Homebrew/shims/shared/git --version
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libarchive.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libb2.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/lz4.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/xz.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/zstd.rb
/usr/local/Homebrew/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/cmake.rb
/usr/bin/env PATH=/usr/local/opt/libarchive/bin:/usr/local/Homebrew/Library/Homebrew/shims/mac/super:/usr/bin:/bin:/usr/sbin:/sbin bsdtar x -f ~/Library/Caches/Homebrew/downloads/8c353ea6176f8b38777694436e18bb3d707b0586c77e38ea122558f22d80a3dc--Space_Cadet.rar -C /private/tmp/d20230517-56188-rn6evf
```

**Review by MikeMcQuaid — APPROVED** [review]
Thanks so much for your first contribution! Without people like you submitting PRs we couldn't run this project. You rock, @TheMDev!

**Pull request merged**